### PR TITLE
[SYCLomatic] Use experimental extension device_global to migrate global __constant__/__device__ variable

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/test/dpct/device_global/device_global.cu [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+__device__ float arr_b[10] = {1, 2, 3};
+__constant__ int arr_c[10] = {3, 2, 1};
+__constant__ int arr_d[10] = {2};
+__constant__ A arr_e[10];
+__device__ B arr_f[10];
+
+
+// CHECK: int device_func() {
+// CHECK:   arr_a.get()[0] = 1;
+// CHECK:   return arr_a.get()[0] + arr_b.get()[0] + arr_c.get()[0] + arr_d.get()[0] + arr_e.get()[0].data + arr_f.get()[0].data;
```
Comment (ziranzha) [inline clang/test/dpct/device_global/device_global.cu]: `.get()` can be removed when used in array subscript expr

File: clang/test/dpct/device_global/device_global.cu [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// CHECK: sycl::ext::oneapi::experimental::device_global<float[10]> arr_b{1, 2, 3};
+// CHECK: /*
+// CHECK: DPCT1124:{{[0-9]+}}: The constant compile-time initialization for device_globals is supported only when compiling with C++20. You may need to adjust the compile commands.
+// CHECK: */
+// CHECK: static sycl::ext::oneapi::experimental::device_global<int[10]> arr_c{3, 2, 1};
+// CHECK: /*
+// CHECK: DPCT1124:{{[0-9]+}}: The constant compile-time initialization for device_globals is supported only when compiling with C++20. You may need to adjust the compile commands.
+// CHECK: */
+// CHECK: static sycl::ext::oneapi::experimental::device_global<int[10]> arr_d{2};
+// CHECK: static sycl::ext::oneapi::experimental::device_global<A[10]> arr_e;
```
Comment (ziranzha) [inline clang/test/dpct/device_global/device_global.cu]: Add `const` for the template parameters when migrating `__constant__`.

File: clang/test/dpct/device_global/device_global.cu [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  var_a = 0;
+  var_e.data = 1;
+  *ptr = var_a + var_b + var_c + var_d + var_e.data + var_f.data + device_func();
+}
+
+int main() {
+    int *dev;
+    cudaMallocManaged(&dev, sizeof(int));
+    *dev = 0;
+    kernel<<<1, 1>>>(dev);
```
Comment (ziranzha) [inline clang/test/dpct/device_global/device_global.cu:97]: Should add test for `cudaMemcpy*Symbol*`.

File: clang/lib/DPCT/AnalysisInfo.h [hunk 4]
```diff
@@ -1847,6 +1859,8 @@ class MemVarInfo : public VarInfo {
 
   static std::unordered_map<const DeclStmt *, int> AnonymousTypeDeclStmtMap;
   bool UseHelperFuncFlag = true;
+  bool UseDeviceGlobalFlag = false;
+  std::string InitForDeviceGlobal;
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.h]: Can reuse `InitList`.

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 5]
```diff
@@ -3848,6 +3870,14 @@ void MemVarMap::getArgumentsOrParametersFromMap(ParameterStream &PS,
       VI.second->setUseHelperFuncFlag(false);
       continue;
     }
+    if (DpctGlobalInfo::useExpDeviceGlobal() &&
+        (VI.second->isConstant() || VI.second->isDevice()) &&
+        !VI.second->isManaged() &&
+        !DpctGlobalInfo::IsVarUsedByRuntimeSymbolAPI(VI.second)) {
+      VI.second->setUseHelperFuncFlag(false);
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Suggest removing unnecessary `MemVarInfo` in `DeviceFunctionInfo::buildInfo`.

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 6]
```diff
@@ -3143,6 +3153,16 @@ std::string MemVarInfo::getMemoryType(const std::string &MemoryType,
   if (isUseHelperFunc()) {
     return buildString(MemoryType, "<", VarType->getBaseName(), ", ",
                        VarType->getDimension(), ">");
+  } else if (isUseDeviceGlobal()) {
+    if (VarType->isArray()) {
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.cpp]: The `if` can be removed

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 7]
```diff
@@ -3104,14 +3105,23 @@ std::string MemVarInfo::getMemoryType() {
     requestFeature(HelperFeatureEnum::device_ext);
     static std::string DeviceMemory =
         MapNames::getDpctNamespace() + "global_memory";
+    if (isUseDeviceGlobal()) {
+      DeviceMemory = MapNames::getClNamespace() +
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Refine the logic.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 8]
```diff
@@ -9028,6 +9028,33 @@ void MemVarRefMigrationRule::runRule(const MatchFinder::MatchResult &Result) {
     };
     const auto *Parent = getParentStmt(MemVarRef);
     bool HasTypeCasted = false;
+    auto Set = DpctGlobalInfo::getUseDeviceGlobalVarSet();
+    auto Info = Global.findMemVarInfo(Decl);
+    std::string Key = Info->getFilePath().getCanonicalPath().str() +
+                      std::to_string(Info->getOffset()) + Info->getName();
+    if (Set.find(Key) == Set.end()) {
```
Comment (ziranzha) [inline clang/lib/DPCT/ASTTraversal.cpp]: How about only try to find the var in `cudaMemcpy*Symbol` APIs and set the flag?

**Review by ziranzha — APPROVED** [review]

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 9]
```diff
@@ -434,6 +434,9 @@ DPCT_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::bits<ExperimentalFeatures>), Exp
         false),
              DPCT_OPT_ENUM("non-uniform-groups", int(ExperimentalFeatures::Exp_NonUniformGroups),
         "Experimental extension that allows use of non-uniform groups.\n",
+        false),
+             DPCT_OPT_ENUM("device_global", int(ExperimentalFeatures::Exp_DeviceGlobal),
+        "Experimental extension that allows device memory allocations similar to C++ global variables.\n",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
        "Experimental extension that allows device scoped memory allocations into SYCL that can be accessed within a kernel using syntax similar to C++ global variables.\n",
```

File: clang/lib/DPCT/Diagnostics.inc [hunk 10]
```diff
@@ -280,6 +280,8 @@ DEF_WARNING(MOVE_TYPE_DEFINITION_DEVICE_FUNC, 1125, MEDIUM_LEVEL, "The type \"%0
 DEF_COMMENT(MOVE_TYPE_DEFINITION_DEVICE_FUNC, 1125, MEDIUM_LEVEL, "The type \"{0}\" defined in function \"{1}\" is used as the parameter type in all functions in the call path from the corresponding sycl::handler::parallel_for() to the current function. You may need to adjust the type definition location.")
 DEF_WARNING(MOVE_TYPE_DEFINITION_KERNEL_FUNC, 1126, MEDIUM_LEVEL, "The type \"%0\" defined in function \"%1\" is used as the parameter type in all functions in the call path from the sycl::handler::parallel_for() to the function \"%1\". You may need to adjust the type definition location.")
 DEF_COMMENT(MOVE_TYPE_DEFINITION_KERNEL_FUNC, 1126, MEDIUM_LEVEL, "The type \"{0}\" defined in function \"{1}\" is used as the parameter type in all functions in the call path from the sycl::handler::parallel_for() to the function \"{1}\". You may need to adjust the type definition location.")
+DEF_WARNING(DEVICE_GLOBAL_INIT, 1127, LOW_LEVEL, "The constant compile-time initialization for device_globals is supported only when compiling with C++20. You may need to adjust the compile commands.")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Diagnostics.inc]:
```suggestion
DEF_WARNING(DEVICE_GLOBAL_INIT, 1127, LOW_LEVEL, "The constant compile-time initialization for device_globals is supported when compiling with C++20. You may need to adjust the compile commands.")
```

File: clang/test/dpct/device_global/device_global.cu [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+// CHECK: /*
+// CHECK: DPCT1127:{{[0-9]+}}: The constant compile-time initialization for device_globals is supported only when compiling with C++20. You may need to adjust the compile commands.
+// CHECK: */
+// CHECK: static sycl::ext::oneapi::experimental::device_global<const float> var_d{1.f};
+// CHECK: sycl::ext::oneapi::experimental::device_global<A> var_e;
+// CHECK: static sycl::ext::oneapi::experimental::device_global<const B> var_f;
+// CHECK: dpct::global_memory<int, 0> var_g;
+__device__ int var_a;
+__device__ int var_b = 0;
+__constant__ float var_c = 2.f;
```
Comment (zhimingwang36) [inline clang/test/dpct/device_global/device_global.cu:36]: The option --optimized-migration impacts __constant__ variable migration. Please summarize a FAQ on constant variable migration in different scenarios: 1. By default. 2 --optimized-migration is given. 3. --optimized-migration and --use-experimental-features=device_global are both given.

File: clang/test/dpct/device_global/device_global.cu [hunk 12]
```diff
@@ -0,0 +1,104 @@
+// RUN: dpct --use-experimental-features=device_global -in-root %S -out-root %T/device_global %S/device_global.cu --cuda-include-path="%cuda-path/include" -- -std=c++14 -x cuda --cuda-host-only
```
Comment (zhimingwang36) [inline clang/test/dpct/device_global/device_global.cu:1]: please check if --use-experimental-features=all works for this lit test.

Comment (intwanghao) [reply clang/test/dpct/device_global/device_global.cu:1]: works

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**